File: 56-merge-intervals/merge-intervals.go
import (
    "sort"
)

func merge(intervals [][]int) [][]int {
    if len(intervals) == 1{
        return intervals
    }
    sort.Slice(intervals, func(i, j int) bool {
        return intervals[i][0] < intervals[j][0]
    })

    i, j := 0, 1
    for j < len(intervals){
        if intervals[j][0] <= intervals[i][1]{
            intervals[i][1] = max(intervals[i][1], intervals[j][1])
            
        } else{
            i++
            intervals[i], intervals[j] = intervals[j], intervals[i]
        }
        j++
    }

    return intervals[:i+1]
}

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        """
        [[1,6],[8,10],[15,18],[15,18]]
                        p
                                      i
        if overlap:
            p[1] = max(endtimes)
        else:
            p++
            p = i
        
            
        """
        if len(intervals) == 1:
            return intervals
        
        intervals.sort(key = lambda x : x[0])
        prev = 0
        for i in range(1, len(intervals)):
            if intervals[i][0] <= intervals[prev][1]:
                intervals[prev][1] = max(intervals[prev][1], intervals[i][1])
            else:
                prev += 1
                intervals[prev] = intervals[i]
        return intervals[:prev + 1]
        